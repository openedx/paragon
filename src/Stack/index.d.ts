import * as React from 'react';

export type DirectionVariants = 'horizontal' | 'vertical';

export interface StackProps extends React.HTMLAttributes<HTMLDivElement> {
  children: React.ReactNode;
  direction?: DirectionVariants;
  gap?: number;
  className?: string;
}

declare const Stack: React.ForwardRefExoticComponent<StackProps>;

export default Stack;
