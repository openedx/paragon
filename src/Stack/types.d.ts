import { FC } from 'react';

interface StackTypes extends HTMLAttributes<HTMLDivElement> {
  direction?: typeof DIRECTION_VARIANTS[number];
  gap?: number;
}

declare const Stack: FC<StackTypes>;

export default Stack;
