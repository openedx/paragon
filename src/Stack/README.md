---
title: 'Stack'
type: 'component'
components:
- Stack
categories:
- Layout
status: 'New'
designStatus: 'Done'
devStatus: 'In progress'
notes: |
---

Shorthand helpers that build on top of our flexbox utilities to make component layout faster and easier than ever. Similar to the [Boostrap Stack](https://react-bootstrap.github.io/layout/stack/) component.

## Basic Usage

## Vertical direction

Stacks are vertical by default and stacked items are full-width by default. Watch this pull request to see more details about the [auto stretching behavior](https://github.com/openedx/paragon/pull/1188).

```jsx live
<Stack gap={3}>
  <Button>first button</Button>
  <Button>second button</Button>
  <Button>third button</Button>
</Stack>
```

## Horizontal direction

```jsx live
<Stack direction="horizontal" gap={3}>
  <div className="border p-2">first block</div>
  <div className="border p-2">second block</div>
  <div className="border p-2">third block</div>
</Stack>
```
