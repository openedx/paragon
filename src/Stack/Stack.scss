@import "variables";

.pgn__vstack,
.pgn__hstack {
  display: flex;
  align-self: stretch;
  gap: $stack-gap;

  @each $level, $space in $spacers {
    &.pgn__stack-gap--#{$level} {
      gap: $space;
    }
  }
}

.pgn__vstack {
  flex: 1 1 auto;
  flex-direction: column;
}

.pgn__hstack {
  flex-direction: row;
  align-items: center;
}
