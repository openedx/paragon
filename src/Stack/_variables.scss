$stack-gap:                       0 !default;
