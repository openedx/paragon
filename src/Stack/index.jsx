import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const DIRECTION_VARIANTS = [
  'horizontal',
  'vertical',
];

const Stack = ({
  direction,
  gap,
  children,
  className,
}) => (
  <div
    className={classNames(
      direction === 'horizontal' ? 'pgn__hstack' : 'pgn__vstack',
      gap ? `pgn__stack-gap--${gap}` : '',
      className,
    )}
  >
    {children}
  </div>
);

Stack.propTypes = {
  /** Specifies the content of the `Stack`. */
  children: PropTypes.node.isRequired,
  /** Specifies direction of the children blocks (column/row). */
  direction: PropTypes.oneOf(DIRECTION_VARIANTS),
  /**
   * Specifies inner space between children blocks.
   *
   * Valid values are based on `the spacing classes`:
   * `0, 0.5, ... 6`.
   */
  gap: PropTypes.number,
  /** Specifies an additional `className` to add to the base element. */
  className: PropTypes.string,
};

Stack.defaultProps = {
  direction: 'vertical',
  gap: 0,
  className: undefined,
};

export default Stack;
