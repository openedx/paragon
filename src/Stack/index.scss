.pgn__vstack,
.pgn__hstack {
  display: flex;
  align-self: stretch;
  gap: var(--pgn-size-stack-gap);

  @each $level, $space in $spacers {
    &.pgn__stack-gap--#{$level} {
      gap: $space;
    }
  }
}

.pgn__vstack {
  flex: 1 1 auto;
  flex-direction: column;

  &.pgn__stack-reversed {
    flex-direction: column-reverse;
  }
}

.pgn__hstack {
  flex-direction: row;
  align-items: center;

  &.pgn__stack-reversed {
    flex-direction: row-reverse;
    justify-content: flex-end;
  }
}
