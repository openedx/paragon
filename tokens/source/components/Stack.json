{
  "stack": {
    "gap": { "value": "0", "source": "$stack-gap" }
  }
}
