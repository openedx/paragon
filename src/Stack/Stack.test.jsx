import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import Stack from './index';

describe('<Stack />', () => {
  describe('correct rendering', () => {
    it('renders without props', () => {
      const tree = renderer.create((
        <Stack>content</Stack>
      )).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders with the vertical direction', () => {
      const wrapper = mount(<Stack>Content</Stack>);
      expect(wrapper.find('.pgn__vstack').length).toBeGreaterThan(0);
      wrapper.setProps({ direction: 'vertical' });
      expect(wrapper.find('.pgn__vstack').length).toBeGreaterThan(0);
    });
    it('renders with the horizontal direction', () => {
      const wrapper = mount(<Stack direction="horizontal">Content</Stack>);
      expect(wrapper.find('.pgn__hstack').length).toBeGreaterThan(0);
    });
    it('renders with the correct gap', () => {
      const gap = 3;
      const wrapper = mount(<Stack gap={gap}>Content</Stack>);
      expect(wrapper.find('.pgn__vstack').hasClass(`pgn__stack-gap--${gap}`)).toEqual(true);
    });
    it('renders with the className prop', () => {
      const className = 'className';
      const wrapper = mount(<Stack className={className}>Content</Stack>);
      expect(wrapper.find('.pgn__vstack').hasClass('className')).toEqual(true);
    });
  });
});
