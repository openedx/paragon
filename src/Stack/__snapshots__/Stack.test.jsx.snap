// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stack /> correct rendering renders without props 1`] = `
<div
  className="pgn__vstack"
>
  content
</div>
`;
